### ISSUE TITLE
Shell completions support

### ISSUE THREAD

**Issue by zefr0x** [Byron/gitoxide#877]
### Summary 💡

What I mean is [shell tab completions](https://en.wikipedia.org/wiki/Command-line_completion).

I think those shells should be supported taking priority into account:
1. Bash
2. Fish
3. Zsh
4. Elvish

Then maybe `PowerShell` for windows, but I don't think any one care 🙃.

---

I thought about using [`clap_complete`](https://crates.io/crates/clap_complete) to auto-generate completion scripts, but after a simple check in source code, we may need to make some changes that could have some complexity to generate them during build-time. However, for run-time this can be easily done, although some additional bloat have to be added to the binary to provide an option like `--generate-completions` in the CLI interface. It can be added as a feature for those who want to enable it before building the tool.

We may face a problem with auto-generation, like in `git` we may need dynamic completions, for example: display available branches or list of commits or staged files. `clap_complete` doesn't support dynamic completions as in https://github.com/clap-rs/clap/issues/1232, but initially it may not be important for `gix`.

### Motivation 🔦

It will make using `gix` easy and more productive.

**Comment by kalkin** [comment]
The issue with `clap_complete` is that you can't define your own completion types and compeltion code. It is very limited. You can't e.g. complete any git hashes.

**Comment by zer0-x** [comment]
We have four choices:
1. Use limited `clap_complete` and wait for  https://github.com/clap-rs/clap/issues/1232.
2. Use `clap_complete` to create initial base then add dynamic parts manually.
3. Use `git` shell completions as a base.
4. Build them from bare metal manually.

**Comment by Byron** [comment]
Thanks for bringing this up! Fix `gix`, as a development tool, I personally have no need for completions or if they happen anything fancy in that, but there is nothing wrong with having more than nothing if the maintenance costs are low. Thus, here I clearly prefer `clap_complete` as it's probably fully automatic, and that's what I remember from the time or two I used it.

For `ein`, one day I'd expect it to have the best-possible completions, but it's not quite there yet and it's unclear when that will happen. Maybe by that time, the *Dynamic completion support* is already available for `clap`.

In short, let's stick to what's basically free and live with the shortcomings of `clap_complete`.
